libsrc/iofuncs/man3/im_header.3
.TH IM_HEADER 3 "7 May 2002"
.SH NAME
im_header_int, im_header_double, im_header_string, im_header_get,
im_header_get_type, im_header_map \- read fields from the image header
.SH SYNOPSIS
#include <vips/vips.h>

int im_header_int( IMAGE *image, const char *field, int *out )
.br
int im_header_double( IMAGE *image, const char *field, double *out )
.br
int im_header_string( IMAGE *image, const char *field, char **out )
.br
GType im_header_get_type( IMAGE *im, const char *field );
.br
int im_header_get( IMAGE *im, const char *field, GValue *value_copy );
.br
typedef void *(*im_header_map_fn)( IMAGE *,
  const char *, GValue *, void *, void * );
.br
void *im_header_map( IMAGE *im, 
  im_header_map_fn fn, void *a, void *b );

.SH DESCRIPTION
.B im_header_int(3)
reads the value of an integer header field. These are
"Xsize", "Ysize", "Bands", "Bbits", "BandFmt", "Coding", and "Type", or any
integer meta field.

.B im_header_double(3)
reads the value of the integer header fields. These are
"Xres", and "Yres", or any double meta field.

.B im_header_string(3)
reads the value of the integer header fields. These are
"Hist", and "filename" or any string meta field. 

.B im_header_get_type(3)
returns the GType (eg. G_TYPE_INT) for a field. It returns zero if the field
does not exist. It does not set 
.B im_error(3),
so it's useful for test for a field's existence.

.B im_header_get(3)
fills the GValue with a copy of the field value, if the field exists. The
value should be zeroed but otherwise uninitialised. The value should be unset
once the user has finished with it.

.B im_header_map(3)
maps a function over all header fields, presenting the value of each field as
a GValue. Return non-NULL from the map function to stop iteration early. It
maps over the builtin fields first, then any meta fields.

.SH RETURN VALUE
All functions returns 0 on success and -1 on error.
.SH SEE ALSO
im_close(3), im_open(3).


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/iofuncs/im_header.c
/* im_header_int, im_header_double, im_header_string: output various fields
 * from the VIPS header
 *
 * 9/7/02 JC
 *	- first version
 * 7/6/05
 * 	- now reads meta fields too
 * 	- cleaned up
 *	- added im_header_exists(), im_header_map()
 * 1/8/05
 * 	- now im_header_get_type() and im_header_get() rather than
 * 	  im_header_exists()
 * 4/1/07
 * 	- removed Hist from standard fields ... now a separate function
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Name, offset pair.
 */
typedef struct _HeaderField {
	const char *field;
	glong offset;
} HeaderField;

/* Built in fields and struct offsets.
 */
static HeaderField int_field[] = {
	{ "Xsize", G_STRUCT_OFFSET( IMAGE, Xsize ) },
	{ "Ysize", G_STRUCT_OFFSET( IMAGE, Ysize ) },
	{ "Bands", G_STRUCT_OFFSET( IMAGE, Bands ) },
	{ "Bbits", G_STRUCT_OFFSET( IMAGE, Bbits ) },
	{ "BandFmt", G_STRUCT_OFFSET( IMAGE, BandFmt ) },
	{ "Coding", G_STRUCT_OFFSET( IMAGE, Coding ) },
	{ "Type", G_STRUCT_OFFSET( IMAGE, Type ) },
	{ "Xoffset", G_STRUCT_OFFSET( IMAGE, Xoffset ) },
	{ "Yoffset", G_STRUCT_OFFSET( IMAGE, Yoffset ) }
};

/* These are actually floats :-( how annoying. We report them as doubles for
 * consistency with the im_meta_*() functions.
 */
static HeaderField double_field[] = {
	{ "Xres", G_STRUCT_OFFSET( IMAGE, Xres ) },
	{ "Yres", G_STRUCT_OFFSET( IMAGE, Yres ) }
};

static HeaderField string_field[] = {
	{ "filename", G_STRUCT_OFFSET( IMAGE, filename ) }
};

int
im_header_int( IMAGE *im, const char *field, int *out )
{
	int i;

	for( i = 0; i < IM_NUMBER( int_field ); i++ )
		if( strcmp( field, int_field[i].field ) == 0 ) {
			*out = G_STRUCT_MEMBER( int, im, 
				int_field[i].offset );
			break;
		}

	if( i == IM_NUMBER( int_field ) &&
		im_meta_get_int( im, field, out ) ) {
		im_error( "im_header_int", 
			_( "no such int field \"%s\"" ), field );
		return( -1 );
	}

	return( 0 );
}

int
im_header_double( IMAGE *im, const char *field, double *out )
{
	int i;

	for( i = 0; i < IM_NUMBER( double_field ); i++ )
		if( strcmp( field, double_field[i].field ) == 0 ) {
			*out = G_STRUCT_MEMBER( float, im, 
				double_field[i].offset );
			break;
		}

	if( i == IM_NUMBER( double_field ) &&
		im_meta_get_double( im, field, out ) ) {
		im_error( "im_header_double", 
			_( "no such double field \"%s\"" ), field );
		return( -1 );
	}

	return( 0 );
}

int
im_header_string( IMAGE *im, const char *field, char **out )
{
	int i;

	for( i = 0; i < IM_NUMBER( string_field ); i++ )
		if( strcmp( field, string_field[i].field ) == 0 ) {
			*out = G_STRUCT_MEMBER( char *, im, 
				string_field[i].offset );
			break;
		}

	if( i == IM_NUMBER( string_field ) &&
		im_meta_get_string( im, field, out ) ) {
		im_error( "im_header_string", 
			_( "no such string field \"%s\"" ), field );
		return( -1 );
	}

	return( 0 );
}

GType 
im_header_get_type( IMAGE *im, const char *field )
{
	int i;
	GType type;

	for( i = 0; i < IM_NUMBER( int_field ); i++ )
		if( strcmp( field, int_field[i].field ) == 0 ) 
			return( G_TYPE_INT );
	for( i = 0; i < IM_NUMBER( double_field ); i++ )
		if( strcmp( field, double_field[i].field ) == 0 ) 
			return( G_TYPE_DOUBLE );
	for( i = 0; i < IM_NUMBER( string_field ); i++ )
		if( strcmp( field, string_field[i].field ) == 0 ) 
			return( G_TYPE_STRING );
	if( (type = im_meta_get_type( im, field )) )
		return( type );

	return( 0 );
}

/* Fill value_copy with a copy of the value, -1 on error. value_copy must be 
 * zeroed but uninitialised. User must g_value_unset( value ).
 */
int
im_header_get( IMAGE *im, const char *field, GValue *value_copy )
{
	int i;

	for( i = 0; i < IM_NUMBER( int_field ); i++ ) 
		if( strcmp( field, int_field[i].field ) == 0 ) {
			g_value_init( value_copy, G_TYPE_INT );
			g_value_set_int( value_copy, 
				G_STRUCT_MEMBER( int, im, 
					int_field[i].offset ) );
			return( 0 );
		}

	for( i = 0; i < IM_NUMBER( double_field ); i++ ) 
		if( strcmp( field, double_field[i].field ) == 0 ) {
			g_value_init( value_copy, G_TYPE_DOUBLE );
			g_value_set_double( value_copy, 
				G_STRUCT_MEMBER( float, im, 
					double_field[i].offset ) );
			return( 0 );
		}

	for( i = 0; i < IM_NUMBER( string_field ); i++ ) 
		if( strcmp( field, string_field[i].field ) == 0 ) {
			g_value_init( value_copy, G_TYPE_STRING );
			g_value_set_static_string( value_copy, 
				G_STRUCT_MEMBER( char *, im, 
					string_field[i].offset ) );
			return( 0 );
		}

	if( !im_meta_get( im, field, value_copy ) )
		return( 0 );

	return( -1 );
}

static void *
header_map_fn( Meta *meta, im_header_map_fn fn, void *a )
{
	return( fn( meta->im, meta->field, &meta->value, a ) );
}

void *
im_header_map( IMAGE *im, im_header_map_fn fn, void *a )
{
	int i;
	GValue value = { 0 };
	void *result;

	for( i = 0; i < IM_NUMBER( int_field ); i++ ) {
		im_header_get( im, int_field[i].field, &value );
		result = fn( im, int_field[i].field, &value, a );
		g_value_unset( &value );

		if( result )
			return( result );
	}

	for( i = 0; i < IM_NUMBER( double_field ); i++ ) {
		im_header_get( im, double_field[i].field, &value );
		result = fn( im, double_field[i].field, &value, a );
		g_value_unset( &value );

		if( result )
			return( result );
	}

	for( i = 0; i < IM_NUMBER( string_field ); i++ ) {
		im_header_get( im, string_field[i].field, &value );
		result = fn( im, string_field[i].field, &value, a );
		g_value_unset( &value );

		if( result )
			return( result );
	}

	if( im->Meta_traverse && 
		(result = im_slist_map2( im->Meta_traverse, 
			(VSListMap2Fn) header_map_fn, fn, a )) )
		return( result );

	return( NULL );
}
